File: coverme_test/src/main/java/com/example/coverme_test/grid_draw.java
package com.example.coverme_test;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.util.AttributeSet;
import android.util.Log;
import android.view.MotionEvent;
import android.view.View;

public class grid_draw extends View {

    private int numColumns, numRows;
    private Float cellWidth, cellHeight;
    private Paint blackPaint = new Paint();
    private boolean[][] cellChecked;
    private int backgroundColors;
    private double count;
    private double porcentage;


    public grid_draw(Context context) {
        this(context, null);
    }

    public grid_draw(Context context, AttributeSet attrs) {
        super(context, attrs);
        blackPaint.setStyle(Paint.Style.FILL_AND_STROKE);
    }

    public int getBackgroundColors() {
        return backgroundColors;
    }

    public void setBackgroundColors(int backgroundColors) {
        this.backgroundColors = backgroundColors;
    }

    public void setColor(int color) {
        blackPaint.setColor(color);
    }

    public void setNumColumns(int numColumns) {
        this.numColumns = numColumns;
        calculateDimensions();
    }


    public void setNumRows(int numRows) {
        this.numRows = numRows;
        calculateDimensions();
    }




    public double getPorcentage() {
        return porcentage;
    }


    @Override
    protected void onSizeChanged(int w, int h, int oldw, int oldh) {
        super.onSizeChanged(w, h, oldw, oldh);
        calculateDimensions();
    }

    private void calculateDimensions() {
        if (numColumns < 1 || numRows < 1) {
            return;
        }
        cellWidth = getWidth() / Float.parseFloat(String.valueOf(numColumns));
        cellHeight = getHeight() / Float.parseFloat(String.valueOf(numRows));

        cellChecked = new boolean[numColumns][numRows];

        invalidate();
    }

    @Override
    protected void onDraw(Canvas canvas) {
        canvas.drawColor(getBackgroundColors());

        if (numColumns == 0 || numRows == 0) {
            return;
        }
        int width = getWidth();
        int height = getHeight();
        for (int i = 0; i < numColumns; i++) {
            for (int j = 0; j < numRows; j++) {
                if (cellChecked[i][j]) {

                    canvas.drawRect(i * cellWidth, j * cellHeight,
                            (i + 1) * cellWidth, (j + 1) * cellHeight,
                            blackPaint);
                }
            }
        }

        for (int i = 1; i < numColumns; i++) {
            canvas.drawLine(i * cellWidth, 0, i * cellWidth, height, blackPaint);
        }

        for (int i = 1; i < numRows; i++) {
            canvas.drawLine(0, i * cellHeight, width, i * cellHeight, blackPaint);
        }
    }


    public void onSelect(MotionEvent event) {
        if (event.getAction() == MotionEvent.ACTION_MOVE) {
            int column = (int) (event.getX() / cellWidth);
            int row = (int) (event.getY() / cellHeight);
            if (column >= 0 && column <= numColumns && row >= 0 && row <= numRows && !cellChecked[column][row]) {
                cellChecked[column][row] = true;
                count++;
                this.porcentage = ((count *100)/(numColumns * numRows));


            }
            invalidate();
        }
    }
}

File: coverme_test/src/main/java/com/example/coverme_test/GridDraw.kt
package com.example.coverme_test

import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.util.AttributeSet
import android.util.Log
import android.view.*


class PixelGridView @JvmOverloads constructor(context: Context, attrs: AttributeSet? = null) : View(context, attrs) {


    var numColumns: Int = 0
        set(numColumns) {
            field = numColumns
            calculateDimensions()
        }
    var numRows: Int = 0
        set(numRows) {
            field = numRows
            calculateDimensions()
        }

    var fillColor : Int = 0
    set(fillColor) {
        field = fillColor
        setColor()
    }

    var strokeW : Float = 0F
        set(strokeW) {
            field = strokeW
            setStrokeW()
        }

    var backgrounColor  : Int = 0
    private var count : Int = 0
    var porcenjate : Double = 0.0
    private var cellWidth: Double = 0.0
    private var cellHeight: Double = 0.0
    private val blackPaint = Paint()
    private var cellChecked: Array<BooleanArray>? = null

    init {
        blackPaint.style = Paint.Style.FILL_AND_STROKE
    }

    private fun getPorcentaje() : Double {
        return porcenjate
    }


    private fun setStrokeW(){
        blackPaint.strokeWidth = strokeW
    }

    private fun setColor(){
        blackPaint.color = fillColor
    }

    override fun onSizeChanged(w: Int, h: Int, oldw: Int, oldh: Int) {
        super.onSizeChanged(w, h, oldw, oldh)
        calculateDimensions()
    }

    private fun calculateDimensions() {
        if (this.numColumns < 1 || this.numRows < 1) {
            return
        }

        cellWidth = width / this.numColumns.toDouble()
        cellHeight = height / this.numRows.toDouble()
        cellChecked = Array(this.numColumns) { BooleanArray(this.numRows) }

        invalidate()
    }

    override fun onDraw(canvas: Canvas) {
        canvas.drawColor(backgrounColor)

        if (this.numColumns == 0 || this.numRows == 0) {
            return
        }

        val width = width
        val height = height

        for (i in 0 until this.numColumns) {
            for (j in 0 until this.numRows) {
                if (cellChecked!![i][j]) {

                    canvas.drawRect(
                        (i * cellWidth).toFloat(), (j * cellHeight).toFloat(),
                        ((i + 1) * cellWidth).toFloat(), ((j + 1) * cellHeight).toFloat(),
                        blackPaint
                    )
                }
            }
        }

        for (i in 1 until this.numColumns) {
            canvas.drawLine((i * cellWidth).toFloat(), 0F, (i * cellWidth).toFloat(), height.toFloat(), blackPaint)
        }

        for (i in 1 until this.numRows) {
            canvas.drawLine(0F, (i * cellHeight).toFloat(), width.toFloat(), (i * cellHeight).toFloat(), blackPaint)
        }
    }

    override fun onTouchEvent(event: MotionEvent): Boolean {
        if (event.action == MotionEvent.ACTION_MOVE) {
            val column = (event.x / cellWidth).toInt()
            val row = (event.y / cellHeight).toInt()
            if ( column in 0 until numColumns && row in 0 until numRows && !cellChecked!![column][row] ){
                count++
                cellChecked!![column][row] = true
                porcenjate = ((count.toDouble() * 100)/(numColumns*numRows))
              //   Log.e("Porcentaje", porcenjate.toString())
            }
            invalidate()
        }
        return true
    }


}